### PR TITLE
Validations

### PR DESCRIPTION
**Pull request by mutinhiri**
In this Pr i added validations for models 
Write unit tests for models
Secure app from n+1 problems

### CODE REVIEW

File: app/views/users/index.html.erb [hunk 1]
```diff
@@ -5,21 +5,21 @@
...
+        <% end %>
         <p class='posts-count'>Number of posts : <%= user.posts_counter %></p>
       </div>
 
     </li>
     <%end%>
   </ul>
   <%=link_to "/users/1/posts/new",class: "new-form-btn", :action => 'go' do%>
- <button class="button">New Post</button>
+  <button class="button">New Post</button>
```
Comment (DeliceLydia) [inline app/views/users/index.html.erb:23]: - [x] When I try to create a post am getting an error you are missing `layouts/alert` partial
post

Comment (DeliceLydia) [reply app/views/users/index.html.erb:23]: - [x] Remember to check if your flash messages are working on `posts#create`, `comments#create`, and `likes#create`

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -23,9 +23,10 @@ def create
...
       format.html do
         if new_post.save
-          redirect_to "/users/#{new_post.user.id}/posts/", notice: 'Success!'
+          redirect_to "/users/#{new_post.user.id}/posts/", flash: { notice: 'Success! Post has been created!' }
         else
-          render :new, alert: 'Error!'
+          render :new, flash.now[:error] = "Error, post couldn't be saved!"
+          render action: 'new'
         end
       end
```
Comment (DeliceLydia) [inline app/controllers/posts_controller.rb:31]: - [x] When i try to see all posts am getting an error
fog

Comment (DeliceLydia) [reply app/controllers/posts_controller.rb:31]: - [x] This is happening as well when I try to see a post by ID `http://localhost:3000/users/1/posts/6` I am getting the same error

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
## Hi @mutinhiri, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there! Here are major highlights are:
- Kudos on writing tests
- No linter errors
- Followed git workflow


## Required Changes ♻️
_Check the comments under the review._



### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)

File: app/controllers/posts_controller.rb (hunk 2 above)
Comment (SajjadAhmad14) [inline app/controllers/posts_controller.rb:31]: I get the same error as pointed out by the previous code reviewer but it occurs when a user clicks on a post after he tries to add a comment on a post.
post-show

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -9,11 +9,11 @@ def create
     @comment.user_id = current_user.id
     @comment.post_id = params[:post_id]
```
Comment (SajjadAhmad14) [inline app/controllers/comments_controller.rb:11]: I get an error when I try to add a comment on a post. Kindly fix the error.
comment

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
## Required Changes ♻️
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hi Bunbee,
### Great work on making the changes requested by a previous reviewer 👏 
### You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

## To Improve 🙏

- [ ] Please check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
**_Note: Please do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Comment by mutinhiri** [comment]
Dear code reviewer my system crashed yesterday and i am still in the middle of setting up my new environment. I could not test the app after making changes so i'd appreciate it if there are errors in this code we schedule a call and offer me some help. This is the last trial i have sent

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Bunbee, 

Your project is complete! Great job for making the necessary changes recommended in the previous review. There is nothing else to say other than... it's time to merge it 🍾 :shipit:. Congratulations! 🎉🥂

<br>
Cheers and Happy coding!👏 💻 🍷 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 1 more limited review per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by mutinhiri** [comment]
Thank you @rloterh

**Pull request merged**